### ISSUE TITLE
language-evolution proposal [for discussion]: first-class “atomic”

### ISSUE THREAD

**Issue by Abe149** [p4lang/p4-spec#1108]

**Comment by vgurevich** [comment]
Given the current situation on the ground, I think it would make sense to re-think this. As far as I am aware, most, if not all, practical implementations do things "atomically" by default, without requiring anything extra. In fact, in many cases doing things otherwise would be harder, not easier.

Moreover, given that P4 code describes the processing of a singe packet separately from all others, there is almost nothing one can do if the process is non-atomic -- just reasoning about it is very hard to begin with.

As a result `@atomic` annotation today achieves nothing, except making the code harder to read. I bet it can be removed from all the programs where it is used today without affecting their behavior. 

Therefore, it might be better if we make the atomic behavior the default. If for some reason in some implementation it might be advantageous to use non-atomic behavior, then they can invent their annotations if needed, and then we'll see if that makes sense to standardize them.

**Comment by Abe149** [comment]
Thanks for the conversation and/or explanation, Vladimir.

In that case, perhaps the spec. should be modified with wording to the effect of “all operations in P4 are atomic unless annotated otherwise.  Any annotation indicating that non-atomic operation is permissible may be freely ignored by the implementation, resulting in a compilation that computes the relevant operation atomically anyway“.

I guess we could try to introduce a standardized annotation e.g. “`nonAtomic_permitted`” and see if anybody actually _wants_ it.

**Comment by jafingerhut** [comment]
I do not understand @vgurevich's answer.  According to the language spec, as far as I can tell, if you write code for a v1model register like this:
```
bit<8> tmp;
bit<8> index = 4;
myreg.read(tmp, index);
tmp = tmp + 1;
myreg.write(index, tmp);
```
and there are no `@atomic` annotations anywhere, then two packets can arrive such that the execution order is that both packets execute the `myreg.read` line one after the other, both reading the same value, then both execute the `tmp = tmp + 1` line in parallel, then both execute the `myreg.write` call in some arbitrary order.  The final result is that the contents of that index of the register will be one more than before.

If you make that block of code atomic, then it is a bug if an implementation ends with the final state having the contents be one more than before.

It is true in the BMv2 v1model implementation that all action bodies behave as if they have `@atomic` around their bodies, even if you do not write that annotation, but that is a detail of its implementation.  Mind you, I think it is a _good_ choice, and one I have recommended to some P4 compiler developers working on a P4->FPGA compiler.

`@atomic` as defined in the language spec is, I suspect, able to be used in more general ways than many implementations would be able to support at exactly the "granularity" specified in the code.  It is of course correct to take blocks of code marked `@atomic` and have the compiler increase the quantity of code they include in some back-end step, since that should only limit the possible behaviors more than the original source code does.

Note: If the above does not sound like a clear case in support of, or against, the current `@atomic` annotation, I agree that it isn't.  I don't necessarily think it is the best way to specify this kind of stuff, nor do I necessarily have something better to offer in its place.

**Comment by Abe149** [comment]
@jafingerhut, thanks for your detailed and thoughtful reply.

Given that reply, I humbly re-instate my request/suggestion: "atomic" should be first class, **not** an annotation.  IMO & TTBOMK, annotations are by definition things that the compiler is allowed to ignore, and the code it generates while ignoring that/those annotation[s] is still supposed to be perfectly valid and to correctly represent what the source program`s semantics are supposed to be, according to the relevant language spec.

Given Andy`s explanation, P4 currently has a language defect IMO: "atomic" being a mere annotation makes it ignorable, yet users can write programs wherein their expectation is that a certain block of code will execute atomically, no matter what limitations the compiler and/or its [current] target may or may not have.

IMO the best solution for this is to promote "atomic" to a keyword.  If/when a compiler encounters that keyword attached correctly to a block of code, and the compiler cannot honor the keyword, then it is responsible for aborting compilation.  To proceed anyway would be a violation of the spec., thus putting that compiler out of compliance.  If it seems like I`m being too harsh here, please bear in mind that what Andy described could [AFAIK, validly] be described as a silent-data-corruption bug.

The second-best approach -- but _much_-less-good than the preceding, IMO -- would be to encode the above into the spec., in the sense of warning compiler builders that e.g. "@atomic, unlike other annotations, is mandatory: if you cannot honor it, you are _required_ to abort compilation".  That would be better than nothing, I hope, but IMO is almost _guaranteed_ to be ignored by one or more implementation, and [of course] adding such text to the spec. _now_ would be irrelevant to existing versions of existing implementations.  A new keyword, OTOH, would break old implementations, which IMO in this context would be a _good_ thing: if the compiler doesn`t "understand" "atomic" and it therefore fails to compile my code, I did _not_ get a switch build that contains a silent corruption bug.

TLDR: failure to compile is better than silent corruption, therefore a keyword for atomic needs is better than an annotation for the same.

**Comment by apinski-cavium** [comment]
I am concered about this with respect to the model we will be implementing here at Marvell.
We will have a huge # (~100 or more) of CPUs that are processing packets at a time. So having a lock between the cores except where `@atomic` is used will cause huge performance issues. Especially in the action part. The wording in the spec is perfect for our usage and I don't think we should change it.
Doing things atomicly by default for us is not an option either. as taking a lock in the fast path is not reasonable at all. Our P4 programs are supposed going to be the fast path and then have a slow path on some other cores if needed.

**Comment by vgurevich** [comment]
@apinski-cavium -- I wonder what exact model do you plan to implement. 

Specifically, is there any example where you would **not** want to implement atomic behavior and where you can't detect whether to take the lock or not? 

Remember, that the problem happens only when you access externs that perform some sort of a read-modify-write operation, correct? Counters, Meters, Registers are primary examples. Would you ever want to allow for race conditions in there? So you would probably have to either implement them in a way that doesn't require locks or you will have locks and I suspect you will be able to do it without the help from this annotation.

**Comment by vgurevich** [comment]
@jafingerhut  -- you are absolutely correct, the current language spec allows this to happen.  I simply stated that no current implementation works this way as far as I know and I can't imagine why any future one would do it. 

If you believe that there are reasons for the future implementations to behave that way, meaning that there will be meaningful programs with the code similar to what you've written, but where the `@atomic` annotation will be harmful, then I'll remove my comment.

**Comment by Abe149** [comment]
@apinski-cavium I am not proposing/suggesting "Doing things atomicly by default" at _all_.  What I _am_ proposing/suggesting, in a nutshell:

- get rid of the '@' in "atomic"
- make it so [old] compilers that don`t "understand" "atomic" will fail to compile code that uses it
- make it clear to compiler vendors that "atomic" is a _contract_, not a "maybe" or a "please" or a "do it if it doesn`t cause too much trouble or cost too much".  Do it, or fail to compile.  No silent-corruption bugs due to buggy compilers.

**Comment by Abe149** [comment]
@vgurevich: re _"the current language spec allows this to happen. I simply stated that no current implementation works this way"_: if the spec. allows for random interleavings when "@atomic" is not used on code like as in Andy`s example, then IMO the main/only Open Source P4 simulator ["BMv2", right?] should model this behavior and produce random interleavings for such code whenever it can.  This way, repeated executions of the same buggy P4 code _should_ [eventually, i.e. given enough executions] produce different results, thereby -- I _hope_ -- causing the lightbulb to come on over the programmer`s head indicating that he/she has realized that there is a data race in her/his program.

**Comment by jafingerhut** [comment]
@vgurevich and I had a brief private chat on this topic, and I came up with an example that may have convinced him of the usefulness of `@atomic` and/or something even stronger in the language than an annotation.

Suppose you have a program like this:
```
bit<8> tmp;
bit<8> index1, index2;

// potentially lots of P4 code here

// code block A starts here
myreg.read(tmp, index1);
tmp = ...;   // some code here to calculate value to write back to the register
myreg.write(index1, tmp);
// code block A ends here

// potentially lots more P4 code here

// code block B starts here
myreg.read(tmp, index2);
tmp = ...;   // some code here to calculate value to write back to the register, potentially different than abov
myreg.write(index2, tmp);
// code block B ends here
```

One could argue that a P4 compiler should perhaps automatically infer critical sections, but if so, for a program like the one above, should there be one critical section for code block A, and a separate one for code block B?  Or should the entire span of code starting at A and ending at B be one big critical section?  Or should the programmer get to have some control over that decision?

If the programmer should have some control over that decision, then some way to specify that in the P4 source code, whether it is `@atomic` or something else, seems useful.

**Comment by vgurevich** [comment]
I agree that if the target wants to allow something like what @jafingerhut just mentioned, then we need to have  a special, first-class way to mark those critical sections, which is what @Abe149 is proposing (promoting `@atomic` to a statement).

One thing I want to mention, though, is that this kind of targets will probably call for many other things that we do not currently support, such as the ability to apply the same table multiple times per packet, perhaps allowing loops, etc... :)

**Comment by Abe149** [comment]
@jafingerhut: I respectfully — but **_strongly_** — disagree with “_a P4 compiler should_ [perhaps] _automatically infer critical sections_”.  IMO & AFAIK, automatic detection of where critical sections begin and end is _far_ from trivial; this is not e.g. dead-code elimination we are talking about here.  IMO, a/the P4 programmer should be responsible for delimiting critical sections appropriately, as needed.  If/when this is not done and it needed/needs to be done, then the data races that result[ed] are/were the programmer｀s fault, not the compiler｀s.

My POV in _general_ is that a compiler｀s job is to translate correct code to correct code; it｀s not a compiler｀s job to fix incorrect code that shows up in the compiler｀s input[s].  IMO critical sections that are not marked as being such are _wrong_.

**Comment by mbudiu-vmw** [comment]
> One thing I want to mention, though, is that this kind of targets will probably call for many other things that we do not currently support, such as the ability to apply the same table multiple times per packet, perhaps allowing loops, etc... :)

Targets such as ebpf or open flow already allow applying a table multiple times. The language allows it

**Comment by jafingerhut** [comment]
This is just a quick data point demonstrating that it seems to me like today's p4c is probably unsafe with regards to moving code from inside of a block annotated with `@atomic`, to outside of that block.  If so, that does leave open the possibility of changing this behavior so it no longer occurs, or introducing some other language construct that is safer in this regard.  I would guess that the LocalCopyPropagation pass is an especially suspicious source of issues of moving code from inside of some block until after that block, whether it has an `@atomic` annotation, or inside of some other new custom code block, unless changes are made to LocalCopyPropagation to prevent such code movement across specific kinds of boundaries.

See the source code, README (copied below), and compiler output files in the attached zip file.

Version of p4c source code that I tested with, although I believe that
the behavior described below has been like this for a long time.

```bash
$ git log -n 1
commit 787e718b28203719662a853a8476a6a4e949ee9d (HEAD -> main, origin/main, origin/HEAD)
Author: Fabian Ruffy <5960321+fruffy@users.noreply.github.com>
Date:   Mon Apr 3 14:36:33 2023 -0400

    Remove the inja submodule and replace it with a CMake fetchcontent instrumentation. (#3959)
```

Compilation command:

```bash
$ mkdir -p dumpdir
$ p4test --dump dumpdir --top4 FrontEndLast,MidEndLast foo2.p4
```

The original definition of action `ct_table_hit` in the source program
is:

```
    bit<48> tmp1;
    bit<48> tmp2;
    bit<48> tmp3;
    action ct_table_hit (bit<48> p1, bit<48> p2) {
        @atomic {
            drop_packet();
            tmp1 = p1 + p2;
            tmp2 = h1.get_hash({tmp1}) + 5;
        }
        tmp3 = tmp1 + 7;
    }
```

But compare that to the version of the action `ct_table_hit` after the
last pass of the midend (I have omitted some `@name` annotations for
brevity):

```
    bit<48> tmp2_0;
    bit<48> tmp3_0;
    bit<48> tmp;
    action ct_table_hit(bit<48> p1, bit<48> p2) {
        @atomic {
            drop_packet();
            tmp = h1_0.get_hash<tuple_0>((tuple_0){f0 = p1 + p2});
            tmp2_0 = tmp + 48w5;
        }
        tmp3_0 = p1 + p2 + 48w7;
    }
```

Now this might not be exactly a smoking gun of a bug, but look
carefully at the assignment to `tmp3` in the original, vs. the
assignment to `tmp3_0` in the output of the midend.

It seems _very_ suspicious to me that the LocalCopyPropagation pass
moved the evaluation of the expression `p1 + p2` from inside the
`@atomic` annotation until after it.

If p4c can move that kind of code from inside of `@atomic` outside of
`@atomic`, what else might it move?

[p4-spec-issue-1108-example.zip](https://github.com/p4lang/p4-spec/files/11143672/p4-spec-issue-1108-example.zip)

**Comment by Abe149** [comment]
@jafingerhut ...

> ```
>     bit<48> tmp2_0;
>     bit<48> tmp3_0;
>     bit<48> tmp;
>     action ct_table_hit(bit<48> p1, bit<48> p2) {
>         @atomic {
>             drop_packet();
>             tmp = h1_0.get_hash<tuple_0>((tuple_0){f0 = p1 + p2});
>             tmp2_0 = tmp + 48w5;
>         }
>         tmp3_0 = p1 + p2 + 48w7;
>     }
> ```

I'm just curious: if we imagine that there was no "@atomic" such that it was OK to do, how could/would duplicating the "p1+p2" subexpression actually be a _good_ thing?  This looks to me like a bad application of copy propagation, unless I misunderstood something (which is quite possible as I haven't been sleeping well lately so my brain is not running at 100% as I write this reply).

**Comment by jafingerhut** [comment]
I am not sure if this is the reason why the LocalCopyPropagation pass exists, but I believe that pass is sometimes a good thing because it can reduce the number of data dependencies between consecutive assignment statements in an action, which on some hardware architectures makes it possible to implement them, vs. not possible.

**Comment by mbudiu-vmw** [comment]
I would think that this is a bug rather than a problem with the language design.
Even if we introduce a new keyword for atomic, we'll have to retrofit the hundreds of passes we already have to cope with it.
Many passes I wrote will keep around blocks that have annotations. Whether that preserves their semantics is something I haven't really audited properly.

**Comment by Abe149** [comment]
TLDR: I _strongly_ agree with most/all of what Mihai recently posted.


On Thu, Apr 6, 2023 at 2:40 PM Mihai Budiu ***@***.***> wrote:
>
> I would think that this is a bug rather than a problem with the language design.
> Even if we introduce a new keyword for atomic, we'll have to retrofit the hundreds of passes we already have to cope with it.
> Many passes I wrote will keep around blocks that have annotations. Whether that preserves their semantics is something I haven't really audited properly.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.Message ID: ***@***.***>



-- 
— A.

**Comment by jnfoster** [comment]
In the interest of tidying up the set of active issues on the P4 specification repository, I'm marking this as "stalled" and closing it. Of course, we can always re-open it in the future if there is interest in resurrecting it.
